File: lunar_calendar.rb
# lunar_calendar.rb — Ruby версия

require 'date'

def julian_day(year, month, day)
  a = (14 - month) / 12
  y = year + 4800 - a
  m = month + 12 * a - 3
  day + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045
end

def moon_phase(year, month, day)
  known_new_moon = 2451550.26
  target_jd = julian_day(year, month, day) + 0.5
  days_since = target_jd - known_new_moon
  age = days_since % 29.53058867
  age += 29.53058867 if age < 0
  age
end

def phase_name(age)
  if age < 1.0 || age >= 29.0
    "🌑 Новолуние"
  elsif age < 7.0
    "🌒 Молодая Луна"
  elsif age < 8.0
    "🌓 Первая четверть"
  elsif age < 14.0
    "🌔 Растущая Луна"
  elsif age < 15.0
    "🌕 Полнолуние"
  elsif age < 21.0
    "🌖 Убывающая Луна"
  elsif age < 22.0
    "🌗 Последняя четверть"
  else
    "🌘 Старая Луна"
  end
end

def illumination(age)
  angle = 2 * Math::PI * age / 29.53058867
  (1 - Math.cos(angle)) / 2 * 100
end

def draw_moon(age)
  illum = illumination(age)
  size = 10
  lines = []
  (0...size).each do |y|
    line = ""
    (0...size).each do |x|
      dx = x - size/2.0
      dy = y - size/2.0
      if dx*dx + dy*dy <= (size/2.0)**2
        nx = dx / (size/2.0)
        if illum > 50
          line << (nx < 0 ? "██" : "  ")
        else
          line << (nx < 0 ? "  " : "██")
        end
      else
        line << "  "
      end
    end
    lines << line
  end
  lines.join("\n")
end

def main
  if ARGV.length > 0
    begin
      year, month, day = ARGV[0].split('-').map(&:to_i)
    rescue
      puts "Неверный формат даты. Используйте ГГГГ-ММ-ДД"
      exit 1
    end
  else
    now = Date.today
    year, month, day = now.year, now.month, now.day
  end

  age = moon_phase(year, month, day)
  phase = phase_name(age)
  illum = illumination(age)

  puts "\e[36m🌙 Lunar Calendar (Ruby)\e[0m"
  puts "Дата: %04d-%02d-%02d" % [year, month, day]
  puts
  puts "\e[32m#{phase}\e[0m"
  puts "Возраст: %.1f дней" % age
  puts "Освещённость: %.1f%%" % illum
  puts
  puts draw_moon(age)
end

main if __FILE__ == $0

File: lunar_calendar.js
// lunar_calendar.js — JavaScript версия

const readline = require('readline');

function julianDay(year, month, day) {
    let a = Math.floor((14 - month) / 12);
    let y = year + 4800 - a;
    let m = month + 12 * a - 3;
    return day + Math.floor((153 * m + 2) / 5) + 365 * y + Math.floor(y / 4) - Math.floor(y / 100) + Math.floor(y / 400) - 32045;
}

function moonPhase(year, month, day) {
    const knownNewMoon = 2451550.26;
    let targetJD = julianDay(year, month, day) + 0.5;
    let daysSince = targetJD - knownNewMoon;
    let age = daysSince % 29.53058867;
    if (age < 0) age += 29.53058867;
    return age;
}

function phaseName(age) {
    if (age < 1.0 || age >= 29.0) return "🌑 Новолуние";
    if (age < 7.0) return "🌒 Молодая Луна";
    if (age < 8.0) return "🌓 Первая четверть";
    if (age < 14.0) return "🌔 Растущая Луна";
    if (age < 15.0) return "🌕 Полнолуние";
    if (age < 21.0) return "🌖 Убывающая Луна";
    if (age < 22.0) return "🌗 Последняя четверть";
    return "🌘 Старая Луна";
}

function illumination(age) {
    let angle = 2 * Math.PI * age / 29.53058867;
    return (1 - Math.cos(angle)) / 2 * 100;
}

function drawMoon(age) {
    let illum = illumination(age);
    let size = 10;
    let lines = [];
    for (let y = 0; y < size; y++) {
        let line = '';
        for (let x = 0; x < size; x++) {
            let dx = x - size/2;
            let dy = y - size/2;
            if (dx*dx + dy*dy <= (size/2)*(size/2)) {
                let nx = dx / (size/2);
                if (illum > 50) {
                    line += nx < 0 ? '██' : '  ';
                } else {
                    line += nx < 0 ? '  ' : '██';
                }
            } else {
                line += '  ';
            }
        }
        lines.push(line);
    }
    return lines.join('\n');
}

function main() {
    const args = process.argv.slice(2);
    let year, month, day;
    if (args.length > 0) {
        let parts = args[0].split('-');
        if (parts.length === 3) {
            year = parseInt(parts[0]);
            month = parseInt(parts[1]);
            day = parseInt(parts[2]);
        } else {
            console.log('Неверный формат даты. Используйте ГГГГ-ММ-ДД');
            process.exit(1);
        }
    } else {
        let now = new Date();
        year = now.getFullYear();
        month = now.getMonth() + 1;
        day = now.getDate();
    }

    let age = moonPhase(year, month, day);
    let phase = phaseName(age);
    let illum = illumination(age);

    console.log('\x1b[36m🌙 Lunar Calendar (JavaScript)\x1b[0m');
    console.log(`Дата: ${year}-${String(month).padStart(2,'0')}-${String(day).padStart(2,'0')}`);
    console.log();
    console.log(`\x1b[32m${phase}\x1b[0m`);
    console.log(`Возраст: ${age.toFixed(1)} дней`);
    console.log(`Освещённость: ${illum.toFixed(1)}%`);
    console.log();
    console.log(drawMoon(age));
}

if (require.main === module) main();
